### PR TITLE
CW-611-Refactor-Address-Handling

### PR DESCRIPTION
**Pull request by MrCyjaneK**
# Description

See jira ticket for details

# Pull Request - Checklist  

- [ ] This PR contains https://github.com/cake-tech/cake_wallet/pull/1620
- [ ] Initial Manual Tests Passed
- [ ] Double check modified code and verify it with the feature/task requirements
- [ ] Format code
- [ ] Look for code duplication
- [ ] Clear naming for variables and methods

### CODE REVIEW

File: cw_monero/lib/monero_wallet_addresses.dart [hunk 1]
```diff
@@ -27,6 +29,40 @@ abstract class MoneroWalletAddressesBase extends WalletAddresses with Store {
   @observable
   String address;
 
+  @override
+  String get latestAddress {
+    final accountIndex = 0;
```
Comment (OmarHatem28) [inline cw_monero/lib/monero_wallet_addresses.dart:34]: shouldn't we be using the index of the account that the user is on?

File: cw_monero/lib/monero_wallet_addresses.dart [hunk 2]
```diff
@@ -27,6 +29,40 @@ abstract class MoneroWalletAddressesBase extends WalletAddresses with Store {
...
+    }
+    return address;
+  }
+
+  @override
+  String get addressForExchange {
+    final accountIndex = 0;
+    var addressIndex = subaddress_list.numSubaddresses(accountIndex) - 1;
+    var address = getAddress(accountIndex: accountIndex,addressIndex: addressIndex);
+    while (hiddenAddresses.contains(address) || manualAddresses.contains(address)) {
```
Comment (OmarHatem28) [inline cw_monero/lib/monero_wallet_addresses.dart:49]: I believe both `latestAddress` and `addressForExchange` should check against both `hiddenAddresses` and `manualAddresses` so there shouldn't be a difference here and they can be merged into one getter `latestUnusedAddress` for example

File: cw_monero/lib/monero_wallet_addresses.dart [hunk 3]
```diff
@@ -27,6 +29,40 @@ abstract class MoneroWalletAddressesBase extends WalletAddresses with Store {
...
+      addressIndex++;
+      address = getAddress(accountIndex: 0, addressIndex: addressIndex);
+    }
+    return address;
+  }
+
+  String getLastAddress(int accountIndex) {
+    var addressIndex = subaddress_list.numSubaddresses(accountIndex) - 1;
+    var address = getAddress(accountIndex: accountIndex,addressIndex: addressIndex);
+    while (hiddenAddresses.contains(address)) {
```
Comment (OmarHatem28) [inline cw_monero/lib/monero_wallet_addresses.dart:59]: maybe I am misunderstanding what the lastAddress is all about, cuz you seem to be skipping the check for manualAddresses every time for the `lastAddress`

**Review by OmarHatem28 — CHANGES_REQUESTED** [review]

File: cw_monero/lib/api/transaction_history.dart [hunk 4]
```diff
@@ -27,7 +33,6 @@ List<Transaction> getAllTransactions() {
   List<Transaction> dummyTxs = [];
 
   txhistory ??= monero.Wallet_history(wptr!);
-  monero.TransactionHistory_refresh(txhistory!);
```
[the remark this answers is not in the document]
Comment (MrCyjaneK) [reply cw_monero/lib/api/transaction_history.dart:30]: To not call it multiple times, it will be called only in one function with a mutex to prevent the crash that occurs when the wallet is large, and refresh is happening on a separate thread

File: lib/monero/cw_monero.dart [hunk 5]
```diff
@@ -61,7 +61,13 @@ class CWMoneroSubaddressList extends MoneroSubaddressList {
   ObservableList<Subaddress> get subaddresses {
     final moneroWallet = _wallet as MoneroWallet;
     final subAddresses = moneroWallet.walletAddresses.subaddressList.subaddresses
-        .map((sub) => Subaddress(id: sub.id, address: sub.address, label: sub.label))
+        .map((sub) => Subaddress(
+          id: sub.id,
+          address: sub.address,
+          label: sub.label,
+          received: sub.balance??"unknown",
```
[the remark this answers is not in the document]
Comment (MrCyjaneK) [reply lib/monero/cw_monero.dart:68]: The UI accepts string, and it is already formatted before passing - otherwise we would have to calculate the string on the UI insteaed of the wallet code

File: cw_monero/lib/monero_wallet_addresses.dart (hunk 3 above)
Comment (MrCyjaneK) [reply cw_monero/lib/monero_wallet_addresses.dart:59]: Correct, when user creates a new address I assume that they want to use it - so I don't exclude them from the list

File: lib/monero/cw_monero.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (MrCyjaneK) [reply lib/monero/cw_monero.dart:68]: received is a string, to account for formating so this is expected.

File: lib/src/screens/receive/receive_page.dart
[the remark this answers is not in the document]
Comment (MrCyjaneK) [reply lib/src/screens/receive/receive_page.dart:1]: ShareUtil was a mistake, it shouldn't be latest - fixed in next commit.

Yeah, it will be better to just reset it when users enters the receive page

File: cw_monero/lib/api/wallet.dart [hunk 6]
```diff
@@ -66,9 +66,14 @@ String getSeedLegacy(String? language) {
   return legacy;
 }
 
-String getAddress({int accountIndex = 0, int addressIndex = 0}) =>
-    monero.Wallet_address(wptr!,
+String getAddress({int accountIndex = 0, int addressIndex = 0}) {
+  while (monero.Wallet_numSubaddresses(wptr!, accountIndex: accountIndex)-1 < addressIndex) {
+    print("adding subaddress");
+    monero.Wallet_addSubaddress(wptr!, accountIndex: accountIndex);
+  }
```
[the remark this answers is not in the document]
Comment (MrCyjaneK) [reply cw_monero/lib/api/wallet.dart:73]: in case when the local state will be different than what's in the monero code, or when we hide excahnge address - then we will just do +1 to the index, and this function should make sure that the address exist (it would return it, but changes to for example label would cause errors)

File: cw_monero/lib/monero_wallet_addresses.dart (hunk 2 above)
Comment (MrCyjaneK) [reply cw_monero/lib/monero_wallet_addresses.dart:49]: We shouldn't use manual addresses for exchange refunds, that's why there are 2 getters, we want to display manual addresses, but we don't want to use them for exchanges.

File: cw_monero/lib/monero_wallet_addresses.dart [hunk 7]
```diff
@@ -27,6 +29,30 @@ abstract class MoneroWalletAddressesBase extends WalletAddresses with Store {
   @observable
   String address;
 
+  @override
+  String get latestAddress {
+    var addressIndex = subaddress_list.numSubaddresses(account?.id??0) - 1;
+    var address = getAddress(accountIndex: account?.id??0, addressIndex: addressIndex);
+    while (hiddenAddresses.contains(address)) {
+      addressIndex++;
+      address = getAddress(accountIndex: 0, addressIndex: addressIndex);
```
Comment (OmarHatem28) [inline cw_monero/lib/monero_wallet_addresses.dart:38]:
```suggestion
      address = getAddress(accountIndex: account?.id??0, addressIndex: addressIndex);
```

File: cw_monero/lib/monero_wallet_addresses.dart [hunk 8]
```diff
@@ -27,6 +29,30 @@ abstract class MoneroWalletAddressesBase extends WalletAddresses with Store {
...
+    return address;
+  }
+
+  @override
+  String get addressForExchange {
+    var addressIndex = subaddress_list.numSubaddresses(account?.id??0) - 1;
+    var address = getAddress(accountIndex: account?.id??0, addressIndex: addressIndex);
+    while (hiddenAddresses.contains(address) || manualAddresses.contains(address) || subaddress_list.getRawLabel(accountIndex: account?.id??0, addressIndex: addressIndex).isNotEmpty) {
+      addressIndex++;
+      address = getAddress(accountIndex: 0, addressIndex: addressIndex);
```
Comment (OmarHatem28) [inline cw_monero/lib/monero_wallet_addresses.dart:50]:
```suggestion
      address = getAddress(accountIndex: account?.id??0, addressIndex: addressIndex);
```

File: cw_monero/lib/monero_wallet_addresses.dart [hunk 9]
```diff
@@ -109,7 +136,10 @@ abstract class MoneroWalletAddressesBase extends WalletAddresses with Store {
         accountIndex: accountIndex,
         defaultLabel: defaultLabel,
         usedAddresses: usedAddresses.toList());
-    subaddress = (subaddressList.subaddresses.isEmpty) ? Subaddress(id: 0, address: address, label: defaultLabel) : subaddressList.subaddresses.last;
+    subaddress = (subaddressList.subaddresses.isEmpty) ? Subaddress(id: 0, address: address, label: defaultLabel, balance: '0', txCount: 0) : subaddressList.subaddresses.last;
```
Comment (OmarHatem28) [inline cw_monero/lib/monero_wallet_addresses.dart:139]: `allAddresses` and `getAllUnusedAddresses` are reversed, let's check if the list is also affected or not (i.e should it be `.first` or keep it as is `.last`)

Comment (MrCyjaneK) [reply cw_monero/lib/monero_wallet_addresses.dart:139]: It is a normal ObservableList<> so shouldn't be reversed - keeping it last makes sense.

Comment (MrCyjaneK) [reply cw_monero/lib/monero_wallet_addresses.dart:139]: https://github.com/cake-tech/cake_wallet/blob/b6089a2dd8d57533f5d7252333c11bb49c9dd466/cw_monero/lib/monero_subaddress_list.dart#L64-L65

File: cw_wownero/lib/api/transaction_history.dart [hunk 10]
```diff
@@ -252,15 +253,13 @@ Future<PendingTransactionDescription> createTransactionMultDest(
 
 class Transaction {
   final String displayLabel;
-  String get subaddressLabel => wownero.Wallet_getSubaddressLabel(wptr!, accountIndex: 0, addressIndex: 0);
-  String get address => wownero.Wallet_address(
-    wptr!,
+  late String subaddressLabel = wownero.Wallet_getSubaddressLabel(wptr!, accountIndex: 0, addressIndex: 0);
+  late String address = getAddress(
```
Comment (OmarHatem28) [inline cw_wownero/lib/api/transaction_history.dart:257]: shouldn't this take the accountIndex and addressIndex instead of 0s?

also why did you change the getters to late in some files?
late will throw an exception if it tried to initialize the variable again, and it's not even final, so it won't give any error if someone tries to set its value again, why not leave them as getters?

File: cw_monero/lib/monero_wallet_addresses.dart (hunk 9 above)
Comment (OmarHatem28) [reply cw_monero/lib/monero_wallet_addresses.dart:139]: ok thanks for looking into it

File: cw_wownero/lib/api/transaction_history.dart (hunk 10 above)
Comment (MrCyjaneK) [reply cw_wownero/lib/api/transaction_history.dart]: fixed the 0 and added proper accountIndex/addressIndex variable there. I've made them final as well - the idea is for the code to be future proof - with `late` it won't getn initialized until it is being used - and next uses are from memory instead of constantly calling a function - which should be more performant.

**Review by OmarHatem28 — APPROVED** [review]
overall looks good
some minor things are missing in Wownero, but not blockers, will commit the minor code suggestions I added and wait for some final tests
